### ISSUE TITLE
Islands

### ISSUE THREAD

**Issue by rogiervandergeer** [godatadriven/evol#39]
I've created a pull request for the islands (#37), but we still need to do some work before we can merge this. In this issue I'll explain some of the changes I've made, and I have a few design decisions to make.

Let's start at the beginning; we have a `Population`, which is a collection of `Individuals` on which we can perform a multitude of operations (`evaluate`, `mutate`, etc). The `Population` can also take an `Evolution` object and perform a series of such operations. 

Now we want to introduce _islands_, where the population is split into several subgroups, on each of which we can perform the same operations as on the `Population`. So we add two (or three, see below) operations: `split` and `combine` in which we can split a population in multiple islands and then merge these islands back into a single population.

I have assumed that we will always want the islands to go through the same evolution. I.e. if we split a population in two islands, and then we apply an evolution on those islands, that the evolution will be the same for both islands. Note that we can pass an island-id to the combiner, picker and mutator functions such that these behave differently on each island - but we assume that the order of the steps is the same. 

**Question 1:** is that a reasonable assumption?

If we will always apply an operation to all islands in the same time, then the interface to the islands will be the same as to the `Population`: calling `mutate` on a group of islands will simply call `mutate` on each island. The same goes for all other operations (except `split` and `combine`). Hence a group of islands behaves exactly the same as a `Population`, and we can apply the same `Evolution` on each (as long as we do not use `split` or `combine`). Since the interface to a group of islands is so close to that of a `Population`, it makes sense to name the group of islands an `IslandPopulation`. 

Now we have `Population`, which is a collection of `Individuals`, and `IslandPopulation`, which has the same interface as `Population` (except, again, for `split` and `combine`) and which is a collection of `Populations`. Since `Population` and `IslandPopulation` have mostly the same interface it makes sense for them to be of the same type, but as `IslandPopulation` contains `Population`s, they cannot inherit from each other. Therefore I've created an abstract base class called `PopulationBase` of which both the `Population` and the `IslandPopulation` inherit. (And since `ContestPopulation` inherits from `Population`, also that inherits from `PopulationBase`.) The great thing about this is that `IslandPopulation` no longer has to be a collection of `Population` objects per se, but can be a collection of `PopulationBase` objects - making it a collection of any type of population.

The above means that operations like `mutate` and `evaluate` can be performed on any object of type `PopulationBase`. But I've had to make exceptions for `split` and `combine`, since these transform a `Population` into an `IslandPopulation` and vice versa. For example, calling `split` on a `Population` should clearly result in an `IslandPopulation`, while calling `combine` on an `IslandPopulation` should result in a `Population` (or `ContestPopulation` if it was a collection of those). Naturally, `Population.combine` should not exist (or throw an error, or do nothing). And now comes the difficult part:

**Question 2:** what should `IslandPopulation.split` do? Should it 
1. Refuse to work, just like you cannot `combine` a `Population`. 
2. Pass the `split` to each island, returning now _the same `IslandPopulation`_, containing not `Population`s but `IslandPopulation`s . E.g. `A[1, 2, 3] -> A[1[a, b], 2[c, d], 3[e, f]]`
3. Split itself, such that it returns _a new `IslandPopulation`, containing islands which contain the
original `Population`s. E.g. `A[1, 2, 3, 4] -> q[A[1, 2], B[3, 4]]`
4. Allow both options 2 and 3 using for example a flag `pushdown` or `level`.
5. Split the internal populations such that the original `IslandPopulation` now contains more islands. E.g. `A[1, 2] -> A[1a, 1b, 2a, 2b]`.  

If we choose for one of the options 2-4 we will have to deal with many layers of sub-islands. This makes functions like `combine` much more complicated, since we would have to specify _which layer_ to combine (or always push down as far as possible) and verify that that layer exists. Also, if we allow sub-islands, we need to make sure that the depth of each branch stays the same. If someone manipulates an `IslandPopulation` to look like this: `A[1, 2[c, d]]`, then what do we do if he requests the second layer to be combined?

Next, we need to decide the workings of `split`. Intuitively I find that `split` should split the population in several groups, such that each individuals ends up in only one of them. I noticed that @koaning expects each individual to end up in _each_ group, effectively duplicating the original `Population`. Since these are two fundamentally different operations, I propose we make both `split()` and `duplicate()` available. 

**Question 3:** do you agree?

This does mean that we have to keep track of the original `intended_size` somehow. For example, suppose I have a `Population` of size 100. If I duplicate this into four groups, I end up with an `IslandPopulation` consisting of four islands with 400 individuals **total**. However, if I split it into four groups, I get an `IslandPopulation` consisting of four each with 25 individuals each, and hence 100 total. If we recombine these, we want to obtain a `Population` with `intended_size == 100`. Of course, when we combine the duplicated islands we will have an initial population of size 400, but this should be reduced back to 100 (either immediately or after the first `survive`, `breed` steps).

For the duplicated islands, it is clear that the `intended_size` of each island is 100. For the split islands, since their initial size is 25, this is not so clear. 

**Question 4:** which is the best option:
1. When using `split`, the intended size of each island is a fraction of the original (25 in the example), and when we use `duplicate` the intended size of each island is the same as the original (100 in the example). We keep track of the original intended size of the whole population and use that when we `combine`.
2. When using `split` or `duplicate`, the intended size of the islands is a fraction of the original (25 in the example). In the case of `duplicate`, each island will initially be much larger than intended.
3. Same as 2, but in the case of `duplicate `, we first use `survive` to cut the original population back to 25 before we duplicate.
4. When using `split` or `duplicate`, the intended size of the islands is the same as the original (100). In the case of `split` each island will initially be much smaller than intended.
5. Same as 4, but in the case of `split` we call `breed` immediately after the split such that all populations are of the right size.

Of course we may allow the user to override the intended size at any point. 

@koaning I'd love to hear your opinion

**Comment by koaning** [comment]
I'll be reading this a few times today and I'll add thoughts below. My first reponse to reading this was ... "maybe we do not want an islandpopulation now that i think of it". Feel free to shoot at this idea, as it only just came in my head. 

### thought 1 

This issue really reminds me of a design decision I had in `kadro` which was inspired by how `dplyr` solves grouping in `R`. 

In `pandas` you can do a very similar thing in dataframes with a `.groupby`. A groupby will split the dataframe into seperate bits and will return a `groupedDataFrame`-object. This made sense back then but this design choice had a few downsides: 

- the `groupedDataFrame`suddenly had methods that the `DataFrame` did not have and vice versa
- the `groupedDataFrame` was designed for aggregation while some people wanted to apply a group such that they could sort within a group, you had to hack a lot to get this to work back in the day.  

R solved this by basically saying "this dataframe has these grouping columns". There is no notion of a different object, it is just that the functions will behave accordingly if there is a grouping active. In the end, you always are dealing with groups in a dataframe; if there is not grouping then the dataframe consists of a single group! 

I am wondering if we need a different population object. We could also have islands be implemented kind of like a `group_by` is implemented in R. After all, a normal population is an island population but on the same island. So why should the normal population be different than the island population?

**Comment by koaning** [comment]
### thought 3 

How would you like to assign populations into islands in practice? I mainly imagine these three:

- randomly 
- iteratively (1 -> 1, 2 -> 2, 3->3, 4->1, 5 -> 2, etc.)
- based on some characteristic in the chromosome/individuals value

But now you want to reassign groups. Maybe you want to exchange members. Basically that means that we're perhaps not exchanging individuals but rather that we are redefining the way we define our groups. So how to you assign members to a new group? 

- randomly 
- iteratively (maybe now 1 -> 1, 2 -> 2, 3->1, 4->2, 5 -> 1, etc.)
- based on some characteristic in the chromosome/individuals value

@rogiervandergeer I'm starting to see benefits of looking at it this way, please feel free to point out the weakness of my approach. It feels like I may be completely glossing over an important aspect.

**Comment by koaning** [comment]
caveat? 

if an islandpopulation has other verbs than the base one ... how are we going to propogate that to the evolution? will we need to check if a verb used in an evolution is available for the population you'd like to apply it on?

**Comment by rogiervandergeer** [comment]
I think this needs to be revived. Part of the discussion took place offline, so I'll summarise what the current status is. Please feel free to point out any things I'm missing. There are several issues that we have not yet decided upon. First, let's start with the functionality; we'll leave the implementation for later.

### Splitting, combining, shuffling

This one should not be too hard. Regardless of the implementation and naming, what operations do we want to implement?

##### A. Splitting

Given a population, split it into islands. We might want to be able to do this:
1. randomly,
2. according to fitness (either skewed or even),
3. according to a user function that takes the chromosome as input,
1. by copying individuals into all islands.

Are there any more options I'm overlooking? Do we want to allow user functions to send a given individual to multiple (or all) islands? I guess 1. and 4. are MVP, the others nice to have.

##### B. Combining

Given islands, combine them into a population.

I don't think there is much to discuss here at this point. We may allow combining islands into larger islands, but I'll leave that discussion for below.

##### C. Shuffling

Given islands, shuffle individuals between them. This can be done randomly, by user defined function, or for example by a predefined pattern.

I think none of this is MVP.

### Sub islands

Again regardless of the implementation, we need to decide whether or not to allow islands within islands. I think we have the following options:

1. Allow sub islands. This is the most generic solution, but rather complicated. When doing _anything_ that has to do with islands, we will have to decide which layer (when combining islands with sub islands, do we combine the islands and keep the sub islands or vice versa?), check whether the islands have the same depth (given islands A and B, where A has sub islands but B does not, what do we do?). 
2. Do not allow sub islands. This is probably the easiest solution, but is a little restricting. A population with islands and a population without islands will have different interfaces (one can split a population without islands, one can combine a population with islands).
3. Mimick sub islands with clever splitting and combining functions. We would allow to split islands into smaller islands without them becoming sub islands, but simply resulting in more islands in the same group of islands. The same would work for combining. Since there would be no hierarchy, keeping track of sub islands will be a tricky task. On the other hand, this is not as restrictive as option 2, and less laborious than option 1.


@koaning, what do you think?

**Comment by koaning** [comment]
## point A 

It feels like we can make a distinction between `.split()` and `.clone()`. I agree that 1 and 4 are the common case. 

## point B 

Combining sure, but an alternative way of looking at it is to say "everyone back to one island". 

## point C 

This is a tricky feature to get right and also something that I can't image you really need. You probably want to do a lot of splitting, combining, splitting, combining, , splitting, combining, etc. 

## point subisland 

I would not allow sub-islands as they can be "mocked" with normal islands. One day we might allow a function to split the population in which case the SubIsland is mocked. Alternatively you could also mock it by having multiple `IslandPopulation`s. This doesn't sound like best practice, but it would still be possible. 

## the big point 

Are we going to have a separate `IslandPopulation`? 

I'll repeat my previous caveats if you're still considering this: 

- if an `IslandPopulation` has other verbs than the base `Population` ... how are we going to propogate that to the evolution? The evolution will no longer be a general object since depending on the population you will be able (or not be able) to use certain verbs. Will we need to check if a verb used in an evolution is available for the population you'd like to apply it on?
- I still observe that our current implementation of `Population` is a special case of the `IslandPopulation` ... namely a one that has only one island.

**Comment by rogiervandergeer** [comment]
So, if we do not allow sub islands, then again regardless of implementation we have the situation that a population without islands (= one island) can be split, and a population with more islands can be combined. So we need to decide what to do when someone calls split on a population with islands and when someone calls combine on a population without islands. (If we were going to allow sub islands, we would only have half this problem). This means that the check you suggest that we need when we use a separate `IslandPopulation` will need to be implemented in any case.

**Comment by koaning** [comment]
My guts is fine with this. It is similar to a `group_by` operation in dataframe-land. You can always remove all groups and you can always redefine how groups are made in a dataframe. But you cannot have groups in groups. 

We really need a face to face (with a whiteboard) discussion about this.

**Comment by rogiervandergeer** [comment]
I agree

**Comment by koaning** [comment]
Should we perhaps close this discussion and make a new one? It's getting rather long.

**Comment by rogiervandergeer** [comment]
No

**Comment by koaning** [comment]
So ... I've started thinking a bit differently about this. There's a principle stuck in my mind. 

> Complicated things need to be possible, but only if simple things remain simple. 

Do we *really* need an island population? One could also just have multiple populations that you mix and match. This way we keep the simple API ... simple.

**Comment by rogiervandergeer** [comment]
We do not _need_ islands. Evol is cool as it is. Anyone who needs islands can build it themselves.

On the other hand: I think it would be cool to have _some_ implementation of islands. This could make for some nice demonstrations (and possibly talks?). It doesn't make sense to implement them outside of Evol, I think. 

I completely agree that we need to keep the API simple. Perhaps we should have another design session to define what the bare minimum of island functionality would be.

**Comment by koaning** [comment]
Cool.
